Content.Shared/GameObjects/Components/Chemistry/SharedSolutionContainerComponent.cs

```diff
@@ -1,115 +1,228 @@
-                {
-                    return;
-                }
+        public ReagentUnit CurrentVolume => Solution.TotalVolume;
 
-                _solution = value;
-                Dirty();
-            }
-        }
+        [ViewVariables]
+        public ReagentUnit EmptyVolume => MaxVolume - CurrentVolume;
 
-        /// <summary>
-        ///     The total volume of all the of the reagents in the container.
-        /// </summary>
         [ViewVariables]
-        public ReagentUnit CurrentVolume => Solution.TotalVolume;
+        public virtual Color Color => Solution.Color;
+
+        [ViewVariables(VVAccess.ReadWrite)]
+        public bool CanReact { get; set; }
 
-        /// <summary>
-        ///     The maximum volume of the container.
-        /// </summary>
         [ViewVariables(VVAccess.ReadWrite)]
-        public ReagentUnit MaxVolume
+        public SolutionContainerCaps Capabilities { get; set; }
+
+        public bool CanExamineContents => Capabilities.HasCap(SolutionContainerCaps.CanExamine);
+
+        public bool CanUseWithChemDispenser => Capabilities.HasCap(SolutionContainerCaps.FitsInDispenser);
+
+        public bool CanAddSolutions => Capabilities.HasCap(SolutionContainerCaps.AddTo);
+
+        public bool CanRemoveSolutions => Capabilities.HasCap(SolutionContainerCaps.RemoveFrom);
+
+        public override void ExposeData(ObjectSerializer serializer)
         {
-            get => _maxVolume;
-            set
-            {
-                if (_maxVolume == value)
-                {
-                    return;
-                }
+            base.ExposeData(serializer);
 
-                _maxVolume = value;
-                Dirty();
-            }
+            serializer.DataField(this, x => x.CanReact, "canReact", true);
+            serializer.DataField(this, x => x.MaxVolume, "maxVol", ReagentUnit.New(0));
+            serializer.DataField(this, x => x.Solution, "contents", new Solution());
+            serializer.DataField(this, x => x.Capabilities, "caps", SolutionContainerCaps.AddTo | SolutionContainerCaps.RemoveFrom | SolutionContainerCaps.CanExamine);
         }
 
-        /// <summary>
-        ///     The current blended color of all the reagents in the container.
-        /// </summary>
-        [ViewVariables(VVAccess.ReadWrite)]
-        public virtual Color SubstanceColor
+        public void RemoveAllSolution()
+        {
+            if (CurrentVolume > 0)
+                return;
+
+            Solution.RemoveAllSolution();
+            ChemicalsRemoved();
+        }
+
+        public bool TryAddReagent(string reagentId, ReagentUnit quantity, out ReagentUnit acceptedQuantity)
+        {
+            acceptedQuantity = EmptyVolume > quantity ? quantity : EmptyVolume;
+            Solution.AddReagent(reagentId, acceptedQuantity);
+
+            if (acceptedQuantity > 0)
+                ChemicalsAdded();
+
+            return acceptedQuantity == quantity;
+        }
```

Comment:
Please add some doc comments to the file. e.g. for this one you'd point out that it will return false for partial reagent insertion.